### PR TITLE
[7.0.0] Experimental extend rule and subrule functionality

### PR DESCRIPTION
**Pull request by comius**
Contains following cherry-picks:

78f8e958b4  hvd@google.com	 Update subrule documentation
b75e322a37  hvd@google.com	 Add support for subrules to depend on other subrules
8473928d8a  ilist@google.com	 Prevent side-effects in initializer callbacks
bdcf7ce2ac  hvd@google.com	 Support Starlark fragments in subrules
804c50305a  hvd@google.com	 Fix bug in subrules that permitted a subrule to call another without declaring a dependency on it
41a1fbc25b  hvd@google.com	 Add a test for subrule bug that permits a subrule to call another
c30826dc83 ilist@google.com	 Rollforward of https://github.com/bazelbuild/bazel/commit/f7946d0107dd75b2f45bcc79b91c016d075a756d: Propagate graph_node_aspect over a computed default attribute
7f2b802542  lberki@google.com   Remove support for loose headers.
358bec8728  ilist@google.com	 Allow setting private attributes from initializers (resolved conflict with ef457ae1430, which was cherry-picked before)
ec46042199  ilist@google.com	 Support initializers on extended rules
eff969d28b  ilist@google.com	 Limit extendability of rules
2d0d22d5b3  hvd@google.com	 Invoke subrule implementations with the corresponding file for single-file attributes
c2dca03ab0  hvd@google.com	 Permit `executable` to be an artifact in `ctx.actions.run()` if it is not from a subrule attribute
567aaf5f67  ilist@google.com	 Check if the parent rule is extendable
2693d9f473  ilist@google.com	 Add an experimental flag for the new API

### CODE REVIEW

**Comment by meteorcloudy** [comment]
Please resolve the conflict ;)

> 7f2b802 [lberki@google.com](mailto:lberki@google.com). Remove support for loose headers.

This commit doesn't exist?

**Comment by comius** [comment]
> Please resolve the conflict ;)

Done

 
> > 7f2b802 [lberki@google.com](mailto:lberki@google.com). Remove support for loose headers.
> 
> This commit doesn't exist?

It exists, just the hash wasn't long enough.

**Comment by meteorcloudy** [comment]
Any idea about the test failures? Are we missing some commits?

**Comment by comius** [comment]
> Any idea about the test failures? Are we missing some commits?

They weren’t there on the first run. @hvadehra mentioned lock file might need to be updated. How do I do that?

**Comment by meteorcloudy** [comment]
The latest failures don't look like lockfile related:
```
java.lang.AssertionError: ERROR <no location>: Traceback (most recent call last):
	File "/virtual_builtins_bzl/common/java/java_common.bzl", line 319, column 32, in <toplevel>
	File "/virtual_builtins_bzl/common/java/java_common.bzl", line 309, column 56, in _make_java_common
Error in _google_legacy_api_enabled: file '@@_builtins//:common/java/java_common.bzl' cannot use private API
```
https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/18470

To update the lockfile, just run `bazel mod deps` locally.

File: src/main/java/com/google/devtools/build/lib/packages/BuiltinRestriction.java [hunk 1]
```diff
@@ -95,12 +93,21 @@ public static void failIfCalledOutsideAllowlist(
...
+  }
+
+  /**
+   * Throws {@code EvalException} if the given {@link Label} is not within either 1) the builtins
+   * repository, or 2) a package or subpackage of an entry in the given allowlist.
+   */
+  public static void failIfLabelOutsideAllowlist(
+      Label label, RepositoryMapping repoMapping, Collection<AllowlistEntry> allowlist)
+      throws EvalException {
+    if (label.getRepository().getNameWithAt().equals("@_builtins")) {
```
Comment (hvadehra) [inline src/main/java/com/google/devtools/build/lib/packages/BuiltinRestriction.java:106]: This seems to have lost the changes from https://github.com/bazelbuild/bazel/commit/c4560822cc9d48f6642fd946ae0eac93a40668cc#diff-b3a4058cfe39033fa0d1ce54e85a94b280c1a34b9e8bb05d3c809ea3496388a2

@comius

Comment (comius) [reply src/main/java/com/google/devtools/build/lib/packages/BuiltinRestriction.java]: Thanks for the hint!

**Pull request merged**